Packs/Phishing/Scripts/PhishingDedupPreprocessingRule/PhishingDedupPreprocessingRule.py

```diff
@@ -0,0 +1,209 @@
+    global MERGED_TEXT_FIELD, FROM_POLICY
+    new_incident_text = new_incident[MERGED_TEXT_FIELD]
+    text = [new_incident_text] + existing_incidents_df[MERGED_TEXT_FIELD].tolist()
+    vectorizer = TfidfVectorizer(token_pattern=r"(?u)\b\w\w+\b|!|\?|\"|\'").fit(text)
+    new_incident_vector = vectorize(new_incident_text, vectorizer)
+    existing_incidents_df['vector'] = existing_incidents_df[MERGED_TEXT_FIELD].apply(lambda x: vectorize(x, vectorizer))
+    existing_incidents_df['similarity'] = existing_incidents_df['vector'].apply(
+        lambda x: cosine_sim(x, new_incident_vector))
+    if FROM_POLICY == FROM_POLICY_DOMAIN:
+        mask = (existing_incidents_df[FROM_DOMAIN_FIELD] != '') &\
+               (existing_incidents_df[FROM_DOMAIN_FIELD] == new_incident[FROM_DOMAIN_FIELD])
+        existing_incidents_df = existing_incidents_df[mask]
+    elif FROM_POLICY == FROM_POLICY_EXACT:
+        mask = (existing_incidents_df[FROM_FIELD] != '') & \
+               (existing_incidents_df[FROM_FIELD] == new_incident[FROM_FIELD])
+        existing_incidents_df = existing_incidents_df[mask]
+    existing_incidents_df.sort_values(by='similarity', ascending=False, inplace=True)
+    if len(existing_incidents_df) > 0:
+        return existing_incidents_df.iloc[0], existing_incidents_df.iloc[0]['similarity']
+    else:
+        return None, None
+
+
+def close_new_incident_and_link_to_existing(new_incident, existing_incident, similarity):
+    entries = []
+    # dumpable_new_incident = {k: v for k, v in new_incident.items() if not isinstance(v, (np.floating, np.integer))}
+    entries.append({'Contents': "Duplicate incident: " + new_incident['name']})
+    entries.append({"Type": entryTypes['note'],
+                    "ContentsFormat": "json",
+                    "Contents": json.dumps(new_incident)
+                    })
+    entries_str = json.dumps(entries)
+    demisto.executeCommand("addEntries", {"id": existing_incident["id"], "entries": entries_str})
+    res = demisto.executeCommand("linkIncidents", {
+        'linkedIncidentIDs': new_incident['id'],
+        'incidentId': existing_incident['id']})
+    if is_error(res):
+        return_error(res)
+    demisto.results('Duplicated incident found: {}, {}'.format(new_incident['id'], existing_incident['id']))
+
+
+def create_new_incident():
+    demisto.results('No duplicated incident found')
+
+
+def main():
+    global EMAIL_BODY_FIELD, EMAIL_SUBJECT_FIELD, EMAIL_HTML_FIELD, FROM_FIELD, MIN_TEXT_LENGTH, FROM_POLICY
+    input_args = demisto.args()
+    EMAIL_BODY_FIELD = input_args.get('emailBody', EMAIL_BODY_FIELD)
+    EMAIL_SUBJECT_FIELD = input_args.get('emailSubject', EMAIL_SUBJECT_FIELD)
+    EMAIL_HTML_FIELD = input_args.get('emailBodyHTML', EMAIL_HTML_FIELD)
+    FROM_FIELD = input_args.get('emailFrom', FROM_FIELD)
+    FROM_POLICY = input_args.get('fromPolicy', FROM_POLICY)
+    existing_incidents = get_existing_incidents(input_args)
```

Comment:
We should have argument to consider closed incidents or not

Reply:
you can add -id:... to filter out current incident - just a suggestion, instead of filter from the dataframe

Reply:
don't want to mess more with the query, it's already complicated (allowing user query + closed/not closed query)

Reply:
ok